### ISSUE TITLE
Don't silently skip missing files on download

### ISSUE THREAD

**Issue by gsfr** [scitran/core#119]
We are currently silently skipping missing or unreadable files on download. As @kofalt points out, this is fundamentally opposed to our data management mission and must be corrected asap.

However, considerable edge cases exist. For example, how do we handle files being legitimately deleted, in parallel to a long-running download operation?

Also, we don't want to inform the user of a (potentially irrelevant) problem 3 hours into a large download and proceed to abort and purge all downloaded data.

We used to check for files during the download preflight, but that was deemed to slow. Let's again explore our options here. This pattern would allow us to mark files as being part of a download and delay physical removal until the download completes.

@kofalt, @ryansanford: Happy to hear your thoughts.

@rentzso: Let's take this on together next week.

**Comment by kofalt** [comment]
:+1: 

So, at the point of download, the browser is fully in control and the user interface, for all intents and purposes, has ceased to exist. I will be pretty adamant that at that stage in the game the only thing we should do is abort on any error.

Before that, are options are wider. Since we currently have this ticket acquisition round trip, I would support making an existence check part of the ticket return, and send back detailed error information on failure. This would allow an interface to alert the user (modal perhaps?) and offer the option to download only the existent data. Choosing that would request a new ticket with fewer files, repeating the process.

For more extensive fiddling, we veer into what basically amounts to disaster recovery, and I'm not currently convinced we need to spend much time building an API for that use case, because at that point somebody should be up to their elbows in the system in question and hands-on diagnosing the issue.

**Comment by rfdougherty** [comment]
Joining this late, so maybe I'm missing something. But in cases of very large downloads, I've seen systems that allow a user to select what they want to download and submit the task to build their download. Then, when the download package is ready, they are notified (e.g., via email) and provided a link to initiate the download. Would that be a viable strategy for downloads of a size that can not be quickly pre-verified?

**Comment by kofalt** [comment]
It definitely could be, but I hope that it won't be necessary in the vast-majority case. I'd be interested to see what (if anything) is slow in an existence check, because that should be very fast.

One pattern I've seen often is where a user interface will show a spinner immediately, then if a process hasn't completed in some short interval it adds additional context - "preparing your download", etc. Something like that should satisfy the 99% use case, I imagine. I would rather not have someone request a tarball of, say, a million files, unless there were no other options; a fuse mount or CLI would probably be a better approach.

**Comment by gsfr** [comment]
I'm actually pretty happy with our on-the-fly tarball generation. It's a nice user experience and also avoids duplication overhead and cleanup. Ultimately, we want to build a system that largely obviates the need for raw data downloads by providing online processing capabilities.

@rentzso did the timing analysis when he added the ability to filter download packages by filetype and other parameters. I don't remember the numbers, though. Conceptually, every time the user makes or changes their selection, a new round trip and file check results. And hitting the filesystem to check for physical file existence is, of course, orders of magnitude slower than simply querying the database.

I could imagine delaying the filesystem check until the last possible moment, that is, the user hitting the download button. And at that time, after confirming the existence of all requested files, we either start streaming or go into an error dialog.

@kofalt: Don't forget that you only have a toy-dataset with grossly truncated scans. The real use-case is multi-gig downloads, containing some hundreds of files. We recently had a request for the entire Simons dataset, which is several hundreds of gigabytes. Clearly indicating to the user what and how much they are about to request is key.

**Comment by rfdougherty** [comment]
I agree that on-the-fly makes for a nice user experience. Apart from the disaster scenarios mentioned (which I agree aren't worth a lot of UI work, since they will necessitate manual intervention), what else could cause 'corrupt' downloads? Some data are deleted mid-stream? A computation is re-run, replacing some results files? Metadata is altered? Permissions are changed? If all these are expected to be rare edge cases, then I think it's fine to abort the download with an error. It may piss off a user on the rare occasion that it happens, but it's a worthy price to pay for assuring data integrity.

And I agree with @kofalt-- we absolutely need a cli method for large data transfers. Even for small transfers, this is much more convenient for the (very common) use-case of a user wanting the data to end up on a remote compute server.

**Comment by kofalt** [comment]
@gsfr 

> The real use-case is multi-gig downloads, containing some hundreds of files.

I don't think the size (in bytes) of an average download changes my position - I'm only proposing an existence check, which is N stat calls. A few thousand should definitely take less than a second. To be clear I'm not proposing we verify their integrity or contents...

> Clearly indicating to the user what and how much they are about to request is key.

That's a good idea. Maybe for bulk downloads we use the ticket-round-trip information I proposed to populate that data. "You are about to download 32 sessions, which is 500 files and ~5 GB."

> I could imagine delaying the filesystem check until the last possible moment

I assert the last possible moment is the ticket request (ref my first comment upthread). The actual download itself is too late - the UI no longer exists nor has any introspection into that request.

@rfdougherty

> what else could cause 'corrupt' downloads?

Good question: I think only an incomplete or malfunctioning filesystem. Files are stored in a [CAS](https://en.wikipedia.org/wiki/Content-addressable_storage) - a custom, in-house CAS that needs some TLC, but still a CAS - and we punt on corruption as the filesystem + sysadmin's problem, currently. I would expect any problems whatsoever to be an extreme edge case.
